server.ts

```diff
@@ -0,0 +1,300 @@
+      }
+    }
+    await fetchPresenceSets(channelsToCheck, true);
+
+    consumeFromAbly();
+  }
+  catch (err: any) {
+    console.log('An error occurred; err = ' + err.toString());
+  }
+}
+
+async function consumeFromAbly() {
+  const endpoint = queueEndpoint;
+  const url = 'amqps://' + apiKey + '@' + endpoint;
+
+  try {
+    // Connect to Ably queue
+    const conn = await Amqp.connect(url);
+    conn.on('error', (err) => { console.error('worker:', 'Connection error!', err); });
+    // Create a communication channel
+    const queueChannel = await conn.createChannel();
+    // Consume from queue
+    const appId = apiKey?.split('.')[0];
+    const presenceQueue = appId + ":" + presenceQueueName;
+    const occupancyQueue = appId + ":" + occupancyQueueName;
+
+    queueChannel.consume(presenceQueue, (item) => {
+      if (!item) {
+        console.warn('queueChannel.consume returned a null item');
+        return;
+      }
+      // Remove message from queue
+      queueChannel.ack(item);
+
+      const decodedEnvelope = JSON.parse(item.content.toString());
+
+      const currentChannel = decodedEnvelope.channel;
+
+      const messages = Ably.Realtime.PresenceMessage.fromEncodedArray(decodedEnvelope.presence);
+
+      messages.forEach((message) => {
+        presenceUpdate(currentChannel, message, false);
+      });
+    });
+
+    queueChannel.consume(occupancyQueue, async (item) => {
+      if (!item) {
+        console.warn('queueChannel.consume returned a null item');
+        return;
+      }
+      // Remove message from queue
+      queueChannel.ack(item);
+
+      const decodedEnvelope = JSON.parse(item.content.toString());
+      const update: OccupancyUpdate = Object.assign(<OccupancyUpdate>{}, decodedEnvelope);
+      for (const occupancy of update.occupancy) {
+        occupancyUpdate(update.channel, occupancy);
+      }
+    });
+  } catch (err) {
+    console.error('worker:', 'Queue error!', err);
+  }
+};
+
+async function fetchPresenceSets(channels: string, updateMemberCount: boolean) {
+  const content = { "channels": channels }
+  // Make a batch request to all relevant channels for their presence sets
+  const presenceSet = await rest.request('GET', '/presence', content);
+  updateCurrentState(presenceSet!.items, updateMemberCount);
+}
+
+function updateCurrentState(presenceSet: string[], updateMemberCount: boolean): void {
+  for (const channelPresenceSet of presenceSet) {
+    const presenceSet = Object.assign(<PresenceSet>{}, channelPresenceSet);
+    const channelName = presenceSet.channel;
+
+    for (const presenceMessage of presenceSet.presence) {
+      presenceUpdate(channelName, presenceMessage, updateMemberCount);
+    }
+  }
+}
+
+function checkPresenceMemberCount(channelId: string): void {
+  if (!storage.has(channelId)) {
+    return;
+  }
+  const channel = storage.get(channelId)!;
+  channel.memberCountCheckTimeout = null;
+  if (channel.expectedPresenceMembers != channel.members.size) {
```

Comment:
I think we'll need to be smarter than this as on a high churn channel this is likely to happen. How about

on an occupancy message being received:
- clear occupancy timer
- store latest occupancy message
- if occupancy and stored presence agree noop
- else we expect to see another presence or occupancy message so set a timer for the max(occupancy interval, presence event connection dying)

on presence message received:
- compare set against last occupancy, if equal clear timer
- else start occupancy timer if it's not already running

on timer fire: 
- no occupancy received when expected, sanity check counts and then resync

This won't spot gaps on high churn channels where occupancy messages are frequent

Reply:
So instead of resetting the occupancy timer on every occupancy and presence update received you only do it on occupancy updates, and for presence ones you just start it if it is not yet started? I'm not sure I get your reasoning behind this.